core/vm/context.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/ethereumproject/go-ethereum/core/state"
+)
+
+// Type is the VM type
+type Type byte
+const (
+	// classic VM
+	ClassicVm Type = iota
+	// sputnik VM
+	SputnikVm
+	// other external VM connected through RPC
+	OtherVm
+)
+
+const (
+	ManageLocalVm = iota
+	UseExternalVm
+)
+const (
+	LocalVm = iota
+	LocalIpcVm
+	LocalRpcVm
+	RemoteRpcVm
+)
+
+const DefaultVm      = ClassicVm
+const DefaultVmProto = LocalVm
+const DefaultVmUsage = ManageLocalVm
+const DefaultVmPort  = 30301
+const DefaultVmHost  = "localhost"
+
+type GasTable struct {
+	ExtcodeSize *big.Int
+	ExtcodeCopy *big.Int
+	Balance     *big.Int
+	SLoad       *big.Int
+	Calls       *big.Int
+	Suicide     *big.Int
+	ExpByte     *big.Int
+
+	// CreateBySuicide occurs when the
+	// refunded account is one that does
+	// not exist. This logic is similar
+	// to call. May be left nil. Nil means
+	// not charged.
+	CreateBySuicide *big.Int
+}
+
+// IsEmpty return true if all values are zero values,
+// which useful for checking JSON-decoded empty state.
+func (g *GasTable) IsEmpty() bool {
+	return reflect.DeepEqual(g, GasTable{})
+}
+
+type Status byte
+
+const (
+	ExitedOk Status = iota
+	Running       // vm is running
+	RequireErr    // vm requires some information
+	TransferErr   // account has insufficient balance and transfer is not possible
+	OutOfGas      // out of gas error occurred
+	BadCode       // bad contract code
+	Terminated    // vm context terminated
+	Broken        // connection with vm is broken or vm is not response
+)
+
+type Fork byte
+
+const (
+	Frontier Fork = iota
+	Homestead
+	Diehard
+)
+
+type Account interface {
+	Address() common.Address
+	Nonce() uint64
+	Balance() *big.Int
+	Code() []byte
+	CodeHash() common.Hash
+}
+
+type RequireID byte
+
+const (
+	RequireAccount RequireID = iota
+	RequireCode
+	RequireHash
+	RequireRules
+)
+
+type Require struct {
+	ID RequireID
+	Address common.Address
+	Number  uint64
+}
+
+type Debugger interface {
+	// Run one instruction and return. If it succeeds, VM status can
+	// still be `Running`.
+	Step() *Require
+}
+
+type Context interface {
+	// Commit an account information to this VM. This should only
+	// be used when receiving `RequireAccountError`.
+	CommitAccount(address common.Address, nonce uint64, balance *big.Int) error
+	// Commit a block hash to this VM. This should only be used when
+	// receiving `RequireBlockhashError`.
+	CommitBlockhash(number uint64,hash common.Hash) error
+	// Commit a contract code to this VM. This should only be used when
+	// receiving `RequireCodeError`.
+	CommitCode(address common.Address, hash common.Hash, code []byte) error
+	// Commit a contract code to this VM. This should only be used when
+	// receiving `RequireRulesError`.
+	CommitRules(number uint64, table *GasTable, fork Fork) error
+	// Finish VM context
+	Finish() error
+	// Returns the changed or committed accounts information up to
+	// current execution status.
+	Accounts() ([]Account, error)
```

Comment:
If you are using the `Account` type defined here, how do you deal with storage changes?